content/blog/gpu-hackathon.md

```diff
@@ -0,0 +1,55 @@
+
+It turns out that this problem is especially bad on a GPU because the only sensible methods for sampling from the binomial distribution involve [rejection sampling](https://en.wikipedia.org/wiki/Rejection_sampling) which is a poor fit for a GPU where you want all your threads ideally doing the same work over and over. By trying to do rejection sampling on a GPU we will get "divergence" where some threads have found their sample, but others are continuing and we're still waiting on them. In order to stop this being totally distracting we also tried implementing a basic volatility model based on brownian motion - involving just draws from a normal distribution we felt this would be a "best case" for our type of stochastic simulation.
+
+With our mentors we started profiling, only to find that we had a major problem with lots of small `cudaMalloc` and `cudaMemcpy` calls rather than doing one big one and dividing that up among the threads. This lead to performance that was considerably slower on the GPU than the CPU and took a day's refactoring to fix.  It also turns out that having lots of small allocations and copies renders the profiler almost unusable so we could not really begin to optimise our "kernel" (the bit of code we were trying to run on the GPU).
+
+With that sorted out, we explored a variety of optimisations for improving our volatility model, focussing on `double` vs `float` maths (with this making a significant difference on the GPU, wheras I'd always thought of `float` as essentially an anachronism on the CPU). We found that as we increased the number of particles (independent simulations) and the number of steps that we ran for, the GPU performance increased from being 5x slower than the speed of the GPU to about 150x faster. It's likely that most of the metrics here are measuring the speed at which we can copy data from R, through a C++ later onto the device, and then reversing this to retrieve the results.
+
+Our SIRS model, which relies on samples from a binomial distribution, was more challenging to optimise, though closer to our real model and with more interesting calculations in the kernel. Our initial version was able to run at best 10x faster than the CPU (even after fixing our memory allocations) which we were a bit disappointed in because running the model on a 10-core system is fairly straightforward.  For the parameters we were using to start optimising, our kernel was taking ~2s to run, vs 1ms for our optimised version of the volatility model!
```

Comment:
```suggestion
Our SIRS model, which relies on samples from a binomial distribution, was more challenging to optimise, though closer to our real model and with more interesting calculations in the kernel. Our initial version was able to run at best 10x faster than the CPU (even after fixing our memory allocations), which we were a bit disappointed in because running the model on a 10-core CPU system is fairly straightforward.  For the parameters we were using to start optimising, our kernel was taking ~2s to run, vs 1ms for our optimised version of the volatility model!
```